src/haz3lcore/statics/TermBase.re

```diff
@@ -243,6 +254,104 @@ and Exp: {
+    | Var(x) => Var(x)
+    | Constructor(x, ty) => Constructor(x, Typ.of_menhir_ast(ty))
+    | Deferral(pos) =>
+      switch (pos) {
+      | InAp => Deferral(InAp)
+      | OutsideAp => Deferral(OutsideAp)
+      }
+    | ListExp(l) => ListLit(List.map(of_menhir_ast, l))
+    | TupleExp(t) =>
+      if (List.length(t) == 1) {
+        Parens(of_menhir_ast(List.hd(t)));
+      } else {
+        Tuple(List.map(of_menhir_ast, t));
+      }
+    | Let(p, e1, e2) =>
+      Let(Pat.of_menhir_ast(p), of_menhir_ast(e1), of_menhir_ast(e2))
+    | FixF(p, e) => FixF(Pat.of_menhir_ast(p), of_menhir_ast(e), None)
+    | TypFun(t, e) => TypFun(TPat.of_menhir_ast(t), of_menhir_ast(e), None)
+    | Undefined => Undefined
+    | TyAlias(tp, ty, e) =>
+      TyAlias(
+        TPat.of_menhir_ast(tp),
+        Typ.of_menhir_ast(ty),
+        of_menhir_ast(e),
+      )
+    | BuiltinFun(s) => BuiltinFun(s)
+    | DeferredAp(f, a) => DeferredAp(of_menhir_ast(f), [of_menhir_ast(a)])
+    | Fun(p, e, name_opt) =>
+      switch (name_opt) {
+      | Some(name_str) =>
+        Fun(
+          Pat.of_menhir_ast(p),
+          of_menhir_ast(e),
+          None,
+          Some(name_str ++ "+"),
+        )
+      | None => Fun(Pat.of_menhir_ast(p), of_menhir_ast(e), None, None)
+      }
+    | ApExp(e1, e2) =>
+      Ap(Operators.Forward, of_menhir_ast(e1), of_menhir_ast(e2))
+    | BinExp(e1, op, e2) =>
+      BinOp(
+        Operators.of_menhir_ast(op),
+        of_menhir_ast(e1),
+        of_menhir_ast(e2),
+      )
+
+    | If(e1, e2, e3) =>
+      If(of_menhir_ast(e1), of_menhir_ast(e2), of_menhir_ast(e3))
+    | CaseExp(e, l) =>
+      let d_scrut = of_menhir_ast(e);
+      let d_rules =
+        List.map(
+          ((pat, exp)) => (Pat.of_menhir_ast(pat), of_menhir_ast(exp)),
+          l,
+        );
+      Match(d_scrut, d_rules);
+    | Cast(e, t1, t2) =>
+      Cast(of_menhir_ast(e), Typ.of_menhir_ast(t1), Typ.of_menhir_ast(t2))
+    | FailedCast(e, t1, t2) =>
+      FailedCast(
+        of_menhir_ast(e),
+        Typ.of_menhir_ast(t1),
+        Typ.of_menhir_ast(t2),
+      )
+    | EmptyHole => EmptyHole
+    | Seq(e1, e2) => Seq(of_menhir_ast(e1), of_menhir_ast(e2))
+    | Test(e) => Test(of_menhir_ast(e))
+    | Cons(e1, e2) => Cons(of_menhir_ast(e1), of_menhir_ast(e2))
+    | ListConcat(e1, e2) =>
+      ListConcat(of_menhir_ast(e1), of_menhir_ast(e2))
+    | Filter(a, cond, body) =>
+      let dcond = of_menhir_ast(cond);
+      let dbody = of_menhir_ast(body);
+      let act = FilterAction.of_menhir_ast(a);
+      Filter(StepperFilterKind.Filter({pat: dcond, act}), dbody);
+    | MultiHole(e) => MultiHole([Exp(of_menhir_ast(e))])
+    | TypAp(e, ty) => TypAp(of_menhir_ast(e), Typ.of_menhir_ast(ty))
+    | UnOp(op, e) =>
+      UnOp(Operators.op_un_of_menhir_ast(op), of_menhir_ast(e))
+    | DynamicErrorHole(e, s) =>
+      DynamicErrorHole(
+        of_menhir_ast(e),
+        InvalidOperationError.t_of_sexp(sexp_of_string(s)),
```

Comment:
Not sure I understand why the sexp encoding is coming into play. Is this just to convert between constant names?

Reply:
Cyrus suggested I encode the error messages in Sexp so that I don't have to create a parser for the error messages.